cmd/notation/key.go

```diff
@@ -71,63 +86,97 @@ var (
 )
 
 func addKey(ctx *cli.Context) error {
-	// initialize
-	args := ctx.Args()
-	switch args.Len() {
-	case 0:
-		return errors.New("missing key and certificate paths")
-	case 1:
-		return errors.New("missing certificate path for the correspoding key")
-	}
-
-	keyPath, err := filepath.Abs(args.Get(0))
-	if err != nil {
-		return err
-	}
-	certPath, err := filepath.Abs(args.Get(1))
+	cfg, err := config.LoadOrDefault()
 	if err != nil {
 		return err
 	}
+	var key *config.KeySuite
+	pluginName := ctx.String("plugin")
 	name := ctx.String("name")
-	if name == "" {
-		name = nameFromPath(keyPath)
-	}
-
-	// check key / cert pair
-	if _, err := tls.LoadX509KeyPair(certPath, keyPath); err != nil {
-		return err
+	if pluginName != "" {
+		key, err = addExternalKey(ctx, pluginName, name)
+	} else {
+		key, err = newX509KeyPair(ctx, name)
 	}
-
-	// core process
-	cfg, err := config.LoadOrDefault()
 	if err != nil {
 		return err
 	}
-	isDefault, err := addKeyCore(cfg, name, keyPath, certPath, ctx.Bool(keyDefaultFlag.Name))
+
+	isDefault := ctx.Bool(keyDefaultFlag.Name)
+	err = addKeyCore(cfg, *key, isDefault)
 	if err != nil {
 		return err
 	}
+
 	if err := cfg.Save(); err != nil {
 		return err
 	}
 
 	// write out
 	if isDefault {
-		fmt.Printf("%s: marked as default\n", name)
+		fmt.Printf("%s: marked as default\n", key.Name)
 	} else {
-		fmt.Println(name)
+		fmt.Println(key.Name)
 	}
 	return nil
 }
 
-func addKeyCore(cfg *config.File, name, keyPath, certPath string, markDefault bool) (bool, error) {
-	if ok := cfg.SigningKeys.Keys.Append(name, keyPath, certPath); !ok {
-		return false, errors.New(name + ": already exists")
+func addExternalKey(ctx *cli.Context, pluginName, keyName string) (*config.KeySuite, error) {
```

Comment:
nit: Will `config.KeySuite` be better than `*config.KeySuite`?

Reply:
Yep.